src/lib.rs

```diff
@@ -701,72 +775,75 @@ impl Drop for ListGuard<'_> {
     }
 }
 
-impl Deref for ListGuard<'_> {
-    type Target = List;
+impl<T> Deref for ListGuard<'_, T> {
+    type Target = List<T>;
 
     #[inline]
-    fn deref(&self) -> &List {
+    fn deref(&self) -> &List<T> {
         &*self.guard
     }
 }
 
-impl DerefMut for ListGuard<'_> {
+impl<T> DerefMut for ListGuard<'_, T> {
     #[inline]
-    fn deref_mut(&mut self) -> &mut List {
+    fn deref_mut(&mut self) -> &mut List<T> {
         &mut *self.guard
     }
 }
 
 /// The state of a listener.
-enum State {
+enum State<T> {
     /// It has just been created.
     Created,
 
     /// It has received a notification.
     ///
     /// The `bool` is `true` if this was an "additional" notification.
-    Notified(bool),
+    Notified(bool, T),
 
     /// An async task is polling it.
     Polling(Waker),
 
     /// A thread is blocked on it.
     Waiting(Thread),
+
+    Dummy,
```

Comment:
I added this `Dummy` state since `Notified` can't be used as a dummy state. `Created` could instead, but I thought having the panics is better security in depth. Let me know if you avoided this intentionally because the branches hurt benchmarks or something like that.